File: pythonProject/src/domain/DirectedGraph.py
class DirectedGraph:
    def __init__(self, vertices: int = 0):
        self._vertices = vertices
        self._edges = 0
        self._outbound = {i: set() for i in range(vertices)}
        self._inbound = {i: set() for i in range(vertices)}
        self._costs = {}

    def _checkVertex(self, vertex: int) -> None:
        if vertex not in self._outbound:
            raise ValueError(f"Invalid vertex: {vertex}")

    def _checkEdge(self, u: int, v: int) -> None:
        self._checkVertex(u)
        self._checkVertex(v)
        if v not in self._outbound[u]:
            raise ValueError(f"Edge {u} -> {v} does not exist")

    def isEdge(self, u: int, v: int) -> bool:
        self._checkVertex(u)
        self._checkVertex(v)
        return v in self._outbound[u]

    def addEdge(self, u: int, v: int, cost: int) -> None:
        self._checkVertex(u)
        self._checkVertex(v)
        if v in self._outbound[u]:
            raise ValueError(f"Edge {u} -> {v} already exists")
        self._outbound[u].add(v)
        self._inbound[v].add(u)
        self._costs[(u, v)] = cost
        self._edges += 1

    def removeEdge(self, u: int, v: int) -> None:
        self._checkEdge(u, v)

        self._outbound[u].remove(v)
        self._inbound[v].remove(u)
        del self._costs[(u, v)]
        self._edges -= 1

    def addVertex(self) -> int:
        newVertex = 0 if not self._outbound else max(self._outbound) + 1
        self._outbound[newVertex] = set()
        self._inbound[newVertex] = set()
        self._vertices += 1
        return newVertex

    def removeVertex(self, vertex: int) -> None:
        self._checkVertex(vertex)

        outboundNeighbors = list(self._outbound[vertex])
        inboundNeighbors = list(self._inbound[vertex])

        for neighbor in outboundNeighbors:
            self.removeEdge(vertex, neighbor)

        for neighbor in inboundNeighbors:
            if neighbor in self._outbound and vertex in self._outbound[neighbor]:
                self.removeEdge(neighbor, vertex)

        del self._outbound[vertex]
        del self._inbound[vertex]
        self._vertices -= 1

    def setEdgeCost(self, u: int, v: int, cost: int) -> None:
        self._checkEdge(u, v)
        self._costs[(u, v)] = cost

    def getEdgeCost(self, u: int, v: int) -> int:
        self._checkEdge(u, v)
        return self._costs[(u, v)]

    def copyGraph(self):
        newGraph = DirectedGraph(0)
        newGraph._vertices = self._vertices
        newGraph._edges = self._edges
        newGraph._outbound = {
            vertex: neighbors.copy()
            for vertex, neighbors in self._outbound.items()
        }
        newGraph._inbound = {
            vertex: neighbors.copy()
            for vertex, neighbors in self._inbound.items()
        }
        newGraph._costs = self._costs.copy()
        return newGraph

    def getVertices(self) -> int:
        return self._vertices

    def getEdges(self) -> int:
        return self._edges

    def getCosts(self) -> {}:
        return self._costs

    def parseVertices(self):
        return sorted(self._outbound.keys())

    def parseOutboundNeighbors(self, vertex: int):
        self._checkVertex(vertex)
        return sorted(self._outbound[vertex])

    def parseInboundNeighbors(self, vertex: int):
        self._checkVertex(vertex)
        return sorted(self._inbound[vertex])

    def getOutDegree(self, vertex: int) -> int:
        self._checkVertex(vertex)
        return len(self._outbound[vertex])

    def getInDegree(self, vertex: int) -> int:
        self._checkVertex(vertex)
        return len(self._inbound[vertex])

File: javaProject/src/domain/DirectedGraph.java
package domain;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.NavigableMap;
import java.util.SortedSet;
import java.util.TreeMap;
import java.util.TreeSet;

public class DirectedGraph {
    private final NavigableMap<Integer, SortedSet<Integer>> inbound;
    private final NavigableMap<Integer, SortedSet<Integer>> outbound;
    private final NavigableMap<Edge, Integer> costs;
    private int vertices;
    private int edges;

    public DirectedGraph(int vertices) {
        if (vertices < 0) {
            throw new IllegalArgumentException("The number of vertices cannot be negative");
        }

        this.inbound = new TreeMap<>();
        this.outbound = new TreeMap<>();
        this.costs = new TreeMap<>();
        this.vertices = vertices;
        this.edges = 0;

        for (int vertex = 0; vertex < vertices; vertex++) {
            outbound.put(vertex, new TreeSet<>());
            inbound.put(vertex, new TreeSet<>());
        }
    }

    public DirectedGraph(DirectedGraph other) {
        this.inbound = new TreeMap<>();
        this.outbound = new TreeMap<>();
        this.costs = new TreeMap<>(other.costs);
        this.vertices = other.vertices;
        this.edges = other.edges;

        for (Map.Entry<Integer, SortedSet<Integer>> entry : other.outbound.entrySet()) {
            outbound.put(entry.getKey(), new TreeSet<>(entry.getValue()));
        }

        for (Map.Entry<Integer, SortedSet<Integer>> entry : other.inbound.entrySet()) {
            inbound.put(entry.getKey(), new TreeSet<>(entry.getValue()));
        }
    }

    private void checkVertex(int vertex) {
        if (!outbound.containsKey(vertex)) {
            throw new IllegalArgumentException("Vertex " + vertex + " not found");
        }
    }

    private void checkEdge(int u, int v) {
        checkVertex(u);
        checkVertex(v);

        if (!outbound.get(u).contains(v)) {
            throw new IllegalArgumentException("Edge " + u + " -> " + v + " not found");
        }
    }

    public boolean isEdge(int u, int v) {
        checkVertex(u);
        checkVertex(v);
        return outbound.get(u).contains(v);
    }

    public void addEdge(int u, int v, int cost) {
        checkVertex(u);
        checkVertex(v);

        if (outbound.get(u).contains(v)) {
            throw new IllegalArgumentException("Edge " + u + " -> " + v + " already exists");
        }

        outbound.get(u).add(v);
        inbound.get(v).add(u);
        costs.put(new Edge(u, v), cost);
        edges++;
    }

    public void removeEdge(int u, int v) {
        checkEdge(u, v);
        outbound.get(u).remove(v);
        inbound.get(v).remove(u);
        costs.remove(new Edge(u, v));
        edges--;
    }

    public int addVertex() {
        int newVertex = outbound.isEmpty() ? 0 : outbound.lastKey() + 1;
        outbound.put(newVertex, new TreeSet<>());
        inbound.put(newVertex, new TreeSet<>());
        vertices++;
        return newVertex;
    }

    public void removeVertex(int vertex) {
        checkVertex(vertex);

        List<Integer> outboundNeighbors = new ArrayList<>(outbound.get(vertex));
        for (int neighbor : outboundNeighbors) {
            removeEdge(vertex, neighbor);
        }

        List<Integer> inboundNeighbors = new ArrayList<>(inbound.get(vertex));
        for (int neighbor : inboundNeighbors) {
            if (outbound.containsKey(neighbor) && outbound.get(neighbor).contains(vertex)) {
                removeEdge(neighbor, vertex);
            }
        }

        outbound.remove(vertex);
        inbound.remove(vertex);
        vertices--;
    }

    public void setEdgeCost(int u, int v, int cost) {
        checkEdge(u, v);
        costs.put(new Edge(u, v), cost);
    }

    public int getEdgeCost(int u, int v) {
        checkEdge(u, v);
        return costs.get(new Edge(u, v));
    }

    public int getVertices() {
        return vertices;
    }

    public int getEdges() {
        return edges;
    }

    public NavigableMap<Edge, Integer> getCosts() {
        return new TreeMap<>(costs);
    }

    public List<Integer> parseVertices() {
        return new ArrayList<>(outbound.navigableKeySet());
    }

    public List<Integer> parseOutboundNeighbors(int vertex) {
        checkVertex(vertex);
        return new ArrayList<>(outbound.get(vertex));
    }

    public List<Integer> parseInboundNeighbors(int vertex) {
        checkVertex(vertex);
        return new ArrayList<>(inbound.get(vertex));
    }

    public int getOutDegree(int vertex) {
        checkVertex(vertex);
        return outbound.get(vertex).size();
    }

    public int getInDegree(int vertex) {
        checkVertex(vertex);
        return inbound.get(vertex).size();
    }

    public DirectedGraph copyGraph() {
        return new DirectedGraph(this);
    }
}
